enn_zoo/enn_zoo/microrts/__init__.py

```diff
@@ -0,0 +1,360 @@
+            jpype.startJVM(convertStrings=False)
+
+        # start microrts client
+        from rts.units import UnitTypeTable
+
+        self.real_utt = UnitTypeTable()
+        from ai.rewardfunction import (
+            RewardFunctionInterface,
+            WinLossRewardFunction,
+            ResourceGatherRewardFunction,
+            AttackRewardFunction,
+            ProduceWorkerRewardFunction,
+            ProduceBuildingRewardFunction,
+            ProduceCombatUnitRewardFunction,
+        )
+
+        self.rfs = JArray(RewardFunctionInterface)(
+            [
+                WinLossRewardFunction(),
+                ResourceGatherRewardFunction(),
+                ProduceWorkerRewardFunction(),
+                ProduceBuildingRewardFunction(),
+                AttackRewardFunction(),
+                ProduceCombatUnitRewardFunction(),
+            ]
+        )
+        self.rfs_names = [str(rf) for rf in self.rfs]
+
+        self.ai2s = [microrts_ai.coacAI for _ in range(1)]
+
+        from ts.entity import JNIEntityClient as Client
+        from ai.core import AI
+
+        self.client = Client(
+            self.rfs,
+            os.path.expanduser(self.microrts_path),
+            self.map_path,
+            self.ai2s[0](self.real_utt),
+            self.real_utt,
+            self.partial_obs,
+        )
+        # get the unit type table
+        self.utt = json.loads(str(self.client.sendUTT()))
+
+    @classmethod
+    def obs_space(cls) -> ObsSpace:
+        return ObsSpace(
+            {
+                "Resource": Entity(["x", "y"]),
+                "Base": Entity(["x", "y"]),
+                "Barracks": Entity(["x", "y"]),
+                "Worker": Entity(["x", "y"]),
+                "Light": Entity(["x", "y"]),
+                "Heavy": Entity(["x", "y"]),
+                "Ranged": Entity(["x", "y"]),
+            }
+        )
+
+    @classmethod
+    def action_space(cls) -> Dict[str, ActionSpace]:
+        return {
+            "unit_action": CategoricalActionSpace(
+                choices=[
+                    "move_up",
+                    "move_right",
+                    "move_down",
+                    "move_left",
+                    "harvest_up",
+                    "harvest_right",
+                    "harvest_down",
+                    "harvest_left",
+                    "return_up",
+                    "return_right",
+                    "return_down",
+                    "return_left",
+                    "produce_base_up",
+                    "produce_base_right",
+                    "produce_base_down",
+                    "produce_base_left",
+                    "produce_barrack_up",
+                    "produce_barrack_right",
+                    "produce_barrack_down",
+                    "produce_barrack_left",
+                ]
+                + [
+                    f"attack_location_{i}" for i in range(49)
```

Comment:
Could be interesting to make this a `SelectEntityAction` which seems like a more natural fit and would allow it to generalize to large maps (if attacking prevents other actions, maybe still have one "attack" choice which determines whether to execute the attack or something else). Probably best left for future PR though, this seems like a good baseline and you're likely to run into a lot of new bugs if you try to actually use `SelectEntityAction`s 😛

Reply:
Yep yep, would be interesting to try it out but probably left for a separate PR.